### PR TITLE
Add a kickass editorconfig for working with CSharp

### PR DESCRIPTION
**Pull request by matiasbeckerle**
## Background

Cool guy @andresmoschini [suggested not long ago](https://github.com/MakingSense/code-style-guides/pull/20#issuecomment-294828897) the possibility of adding an `.editorconfig` specially tuned for working with dotnet. Because is a fantastic idea, here is a PR completely about that.

## Changes

- Added `.editorconfig` based on [Roslyn](https://github.com/dotnet/roslyn/blob/master/.editorconfig)
- Added references in the readme.

## Notes

- There are a couple of newline settings that I had to skip for now because I'm not sure how they work. There is no documentation for now. Look for `csharp_new_line_before_[something]`. If anyone is sure how they work I can add them.
- There are a couple of nice rules that I left there besides not being documented in the readme. Should I remove them? Like the ones about "Suggest more modern language features when available".

### CODE REVIEW

File: CSharp/.editorconfig [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+[*.{csproj,vbproj,vcxproj,vcxproj.filters,proj,projitems,shproj}]
+indent_size = 2
+
+# Xml config files
+[*.{props,targets,ruleset,config,nuspec,resx,vsixmanifest,vsct}]
+indent_size = 2
+
+# JSON files
+[*.json]
+indent_size = 2
```
Comment (AlphaGit) [inline CSharp/.editorconfig]: Based on the standards that we have for JavaScript... should this be 4? (I don't have strong opinions on the matter.)

File: CSharp/.editorconfig [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+# Use language keywords instead of framework type names for type references
+dotnet_style_predefined_type_for_locals_parameters_members = true:suggestion
+dotnet_style_predefined_type_for_member_access = true:suggestion
+
+# Suggest more modern language features when available
+dotnet_style_object_initializer = true:suggestion
+dotnet_style_collection_initializer = true:suggestion
+dotnet_style_coalesce_expression = true:suggestion
+dotnet_style_null_propagation = true:suggestion
+dotnet_style_explicit_tuple_names = true:suggestion
```
Comment (AlphaGit) [inline CSharp/.editorconfig]: You asked if we should include these in the readme. I would say _yes_ but keep in mind that the rule should be "prefer X to Y", always keeping in mind that there might be exceptions that we want to avoid. The purpose of these is to make code more readable, so it really depends on each particular case.

File: CSharp/.editorconfig [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+# CSharp code style settings:
+[*.cs]
```
Comment (AlphaGit) [inline CSharp/.editorconfig]: I think this header might not be needed: it's the same as above.

File: CSharp/.editorconfig [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+# Prefer method-like constructs to have a block body
+csharp_style_expression_bodied_methods = false:none
+csharp_style_expression_bodied_constructors = false:none
+csharp_style_expression_bodied_operators = false:none
+
+# Prefer property-like constructs to have an expression-body
+csharp_style_expression_bodied_properties = true:none
+csharp_style_expression_bodied_indexers = true:none
+csharp_style_expression_bodied_accessors = true:none
```
Comment (AlphaGit) [inline CSharp/.editorconfig]: For these ones and the one for methods -- what is it suggesting exactly? Is it preferring the multiline declaration as opposed to the lambda declaration forms?

File: CSharp/.editorconfig (hunk 1 above)
Comment (matiasbeckerle) [inline CSharp/.editorconfig]: I like 2 for js related and we actually have 2 as standard: https://github.com/MakingSense/code-style-guides/tree/master/Javascript(ES6)#whitespace -- or are you thinking on something else?

File: CSharp/.editorconfig (hunk 2 above)
Comment (matiasbeckerle) [inline CSharp/.editorconfig]: I totally agree with that but... should I remove them then? Should I put a note here? What are you considering as an action item?

File: CSharp/.editorconfig (hunk 3 above)
Comment (matiasbeckerle) [inline CSharp/.editorconfig]: Is slightly the same: the one from above is for "dotnet" and this one is for "C#". It's split like that on the docs. You still want me to remove it?

File: CSharp/.editorconfig (hunk 4 above)
Comment (matiasbeckerle) [inline CSharp/.editorconfig]: - The false ones (methods) is favoring multi-line.
- The truthy ones (properties) is favoring lambda.

**Comment by matiasbeckerle** [comment]
Thanks for the feedback & questions @AlphaGit. Please let me know if I still should change/update something.

**Comment by afantini** [comment]
I'd suggest reviewing the `.editorconfig` files we are using in our projects before closing this PR.

**Comment by matiasbeckerle** [comment]
Thanks @afantini! Send all what you have.

File: CSharp/.editorconfig (hunk 1 above)
Comment (andresmoschini) [inline CSharp/.editorconfig]: +1 for 2 spaces for `*.json`, and by the way also for `*.js`, `*.html` and `*.cshtml`

Comment (matiasbeckerle) [inline CSharp/.editorconfig]: Fair to add, thanks!

Comment (AlphaGit) [inline CSharp/.editorconfig]: @matiasbeckerle Nope, I just got confused. You're right, 2 is fine. Thanks!

File: CSharp/.editorconfig (hunk 2 above)
Comment (AlphaGit) [inline CSharp/.editorconfig]: If these rules are already suggestions, then no modification is required in here. Otherwise, we'd need to change them so that they don't generate errors, only warnings. (Seems the case, if that's what `suggestion` does.)

If these rules are not yet included in our readme file, I would add them with the wording I mentioned.

File: CSharp/.editorconfig (hunk 3 above)
Comment (AlphaGit) [inline CSharp/.editorconfig]: Aren't they the same? Compare line 48 to 28. The comments are ignored but the header is the same. I would remove it and leave the comments as some kind of headers.

Anyway, it's not really important.

File: CSharp/.editorconfig (hunk 4 above)
Comment (AlphaGit) [inline CSharp/.editorconfig]: Awesome, thanks!

File: CSharp/.editorconfig [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+# Xml config files
+[*.{props,targets,ruleset,config,nuspec,resx,vsixmanifest,vsct}]
+indent_size = 2
+
+# JavaScript/HTML
+[*.{js,ts,json,html,cshtml}]
+indent_size = 2
+
+# Dotnet code style settings
```
Comment (mvazquez-ms) [inline CSharp/.editorconfig]: What about adding space preferences?

```
# Space preferences
csharp_space_after_cast = false
csharp_space_after_colon_in_inheritance_clause = true
csharp_space_after_comma = true
csharp_space_after_dot = false
csharp_space_after_keywords_in_control_flow_statements = true
csharp_space_after_semicolon_in_for_statement = true
csharp_space_around_binary_operators = before_and_after
csharp_space_around_declaration_statements = do_not_ignore
csharp_space_before_colon_in_inheritance_clause = true
csharp_space_before_comma = false
csharp_space_before_dot = false
csharp_space_before_open_square_brackets = false
csharp_space_before_semicolon_in_for_statement = false
csharp_space_between_empty_square_brackets = false
csharp_space_between_method_call_empty_parameter_list_parentheses = false
csharp_space_between_method_call_name_and_opening_parenthesis = false
csharp_space_between_method_call_parameter_list_parentheses = false
csharp_space_between_method_declaration_empty_parameter_list_parentheses = false
csharp_space_between_method_declaration_name_and_open_parenthesis = false
csharp_space_between_method_declaration_parameter_list_parentheses = false
csharp_space_between_parentheses = false
csharp_space_between_square_brackets = false
```

Comment (mvazquez-ms) [inline CSharp/.editorconfig]: And

```
dotnet_naming_style.camel_case_underscore_style.required_prefix = _
dotnet_naming_style.camel_case_underscore_style.capitalization = camel_case 
```

File: CSharp/.editorconfig [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+[*.cs]
+
+# Avoid "this." and "Me." if not necessary
+dotnet_style_qualification_for_field = false:suggestion
+dotnet_style_qualification_for_property = false:suggestion
+dotnet_style_qualification_for_method = false:suggestion
+dotnet_style_qualification_for_event = false:suggestion
+
+# Use language keywords instead of framework type names for type references
+dotnet_style_predefined_type_for_locals_parameters_members = true:suggestion
```
Comment (mvazquez-ms) [inline CSharp/.editorconfig]: This could be a warning.

File: CSharp/.editorconfig (hunk 5 above)
Comment (mvazquez-ms) [inline CSharp/.editorconfig]: And

```
# New line preferences
csharp_new_line_before_open_brace = all
csharp_new_line_before_else = true
csharp_new_line_before_catch = true
csharp_new_line_before_finally = true
csharp_new_line_before_members_in_object_initializers = true
csharp_new_line_before_members_in_anonymous_types = true
csharp_new_line_within_query_expression_clauses = true

# Indentation preferences
csharp_indent_block_contents = true
csharp_indent_braces = false
csharp_indent_case_contents = true
csharp_indent_switch_labels = true
csharp_indent_labels = flush_left
```

Comment (mvazquez-ms) [inline CSharp/.editorconfig]: You can review all of these and more on BetterCare's `.editorconfig` (which was in time used for @afantini's project, with slight modifications I believe)

File: CSharp/.editorconfig (hunk 3 above)
Comment (matiasbeckerle) [inline CSharp/.editorconfig]: The target is the same: `.cs` files but the meaning is different: group by dotnet and group by C# so for now I prefer to leave them as is.

File: CSharp/.editorconfig (hunk 5 above)
Comment (matiasbeckerle) [inline CSharp/.editorconfig]: That's cool! Where do you get these? I want to see examples to make sure to set the correct values

File: CSharp/.editorconfig (hunk 2 above)
Comment (matiasbeckerle) [inline CSharp/.editorconfig]: Then I'm going to delete them, if at some point we introduce the styles in the readme, we can add them again.

**Comment by matiasbeckerle** [comment]
This is ready for a second round! I tried to apply the feedback from all of you. Thanks! @andresmoschini @afantini @AlphaGit @mvazquez-ms

**Comment by matiasbeckerle** [comment]
Bump! no comments since some days

@andresmoschini @afantini @AlphaGit @mvazquez-ms

**Comment by AlphaGit** [comment]
@matiasbeckerle Thanks for the bump -- you still have my approval. 👍

**Comment by afantini** [comment]
Sorry I haven't had time to review this yet.
@matiasbeckerle  I just added you as a collaborator on my current project so you can take a look at the `.editorconfig` file.

File: CSharp/.editorconfig [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+csharp_space_before_semicolon_in_for_statement = false:error
+csharp_space_between_empty_square_brackets = false:error
+csharp_space_between_method_call_empty_parameter_list_parentheses = false:error
+csharp_space_between_method_call_name_and_opening_parenthesis = false:error
+csharp_space_between_method_call_parameter_list_parentheses = false:error
+csharp_space_between_method_declaration_empty_parameter_list_parentheses = false:error
+csharp_space_between_method_declaration_name_and_open_parenthesis = false:error
+csharp_space_between_method_declaration_parameter_list_parentheses = false:error
+csharp_space_between_parentheses = false:error
+csharp_space_between_square_brackets = false:error
```
Comment (mvazquez-ms) [inline CSharp/.editorconfig]: Be aware that treating warnings as errors could lead to annoying issues.

E.g.

- NetCore generates `AssemblyInfo.cs` as part of the build process and that file doesn't comply with some standard rules (e.g. it is encoded in UTF-8 without BOM, a rule that by default is a warning in stylecop).  That might not comply with some of our custom rules too and could lead to builds failing for that reason. (happened in my project). The same could happen if one of our rules doesn't not match the standard way that file is generated.

- Auto-generated code. I would add a note to clarify that auto-generated code is not required to comply with these rules but it's hard to exclude specific files from static code analysis. So, for example, on EF7 migrations, you need to manually add `// <auto-generated/>` at the top of the file to avoid having a bunch of warnings/error from auto-generated code, and modifying it manually to comply would be a mistake.

TL;DR; Careful with treating warnings as errors, and add a note about using `// <auto-generated/>` to skip static code analysis for specific files.

This is particularly important when dealing with CodeAnalisys rules, as they could prevent the application from building. The errors here are just a red icon on VS, so.. not a big deal.

**Comment by matiasbeckerle** [comment]
Thank you @afantini, will take a look.

File: CSharp/.editorconfig (hunk 7 above)
Comment (matiasbeckerle) [inline CSharp/.editorconfig]: Ok sounds good @mvazquez-ms. One thing I would like me you to clarify: these ones that I place as errors, should be warnings instead? Because I think I use the same notification level as the StyleCop. Maybe those are wrong too?

Comment (mvazquez-ms) [inline CSharp/.editorconfig]: The difference is that StyleCop errors/warnings will be shown as build output (an error will prevent a project from building successfully). The `.editorConfig` warning/error/suggestions are just yellow/red/green markers next to the code.

e.g. A project containing an EF7 migration will throw build errors with the current StyleCop config. This can be fixed by adding `// <auto-generated/>` at the top of the migration file.

Comment (matiasbeckerle) [inline CSharp/.editorconfig]: 🆒 thanks for clarifying, I understand now.

Comment (matiasbeckerle) [inline CSharp/.editorconfig]: Please review again, tooling section.

**Comment by matiasbeckerle** [comment]
Yay! Finally!
